pkg/planner/core/plan_cache_instance.go

```diff
@@ -0,0 +1,220 @@
+	// Put puts the key and value into the cache.
+	Put(sctx sessionctx.Context, key string, value, opts any) (succ bool)
+	// Evict evicts some cached values.
+	Evict(sctx sessionctx.Context) (evicted bool)
+	// MemUsage returns the total memory usage of this plan cache.
+	MemUsage(sctx sessionctx.Context) int64
+}
+
+// NewInstancePlanCache creates a new instance level plan cache.
+func NewInstancePlanCache(softMemLimit, hardMemLimit int64) InstancePlanCache {
+	planCache := new(instancePlanCache)
+	planCache.softMemLimit.Store(softMemLimit)
+	planCache.hardMemLimit.Store(hardMemLimit)
+	return planCache
+}
+
+type instancePCNode struct {
+	value    *PlanCacheValue
+	lastUsed atomic.Time
+	next     atomic.Pointer[instancePCNode]
+}
+
+// instancePlanCache is a lock-free implementation of InstancePlanCache interface.
+// [key1] --> [headNode1] --> [node1] --> [node2] --> [node3]
+// [key2] --> [headNode2] --> [node4] --> [node5]
+// [key3] --> [headNode3] --> [node6] --> [node7] --> [node8]
+// headNode.value is always empty, headNode is designed to make it easier to implement.
+type instancePlanCache struct {
+	heads   sync.Map
+	totCost atomic.Int64
+
+	evictMutex   sync.Mutex
+	softMemLimit atomic.Int64
+	hardMemLimit atomic.Int64
+}
+
+func (pc *instancePlanCache) getHead(key string, create bool) *instancePCNode {
+	headNode, ok := pc.heads.Load(key)
+	if ok { // cache hit
+		return headNode.(*instancePCNode)
+	}
+	if !create { // cache miss
+		return nil
+	}
+	newHeadNode := pc.createNode(nil)
+	actual, _ := pc.heads.LoadOrStore(key, newHeadNode)
+	if headNode, ok := actual.(*instancePCNode); ok { // for safety
+		return headNode
+	}
+	return nil
+}
+
+// Get gets the cached value according to key and opts.
+func (pc *instancePlanCache) Get(sctx sessionctx.Context, key string, opts any) (value any, ok bool) {
+	headNode := pc.getHead(key, false)
+	if headNode == nil { // cache miss
+		return nil, false
+	}
+	return pc.getPlanFromList(sctx, headNode, opts)
+}
+
+func (*instancePlanCache) getPlanFromList(sctx sessionctx.Context, headNode *instancePCNode, opts any) (any, bool) {
+	for node := headNode.next.Load(); node != nil; node = node.next.Load() {
+		var matchOpts *PlanCacheMatchOpts
+		if opts != nil {
+			matchOpts = opts.(*PlanCacheMatchOpts)
+		}
+		if matchCachedPlan(sctx, node.value, matchOpts) { // v.Plan is read-only, no need to lock
+			node.lastUsed.Store(time.Now()) // atomically update the lastUsed field
+			return node.value, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the key and values into the cache.
+// Due to some thread-safety issues, this Put operation might fail, use the returned succ to indicate it.
+func (pc *instancePlanCache) Put(sctx sessionctx.Context, key string, value, opts any) (succ bool) {
+	vMem := value.(*PlanCacheValue).MemoryUsage()
+	if vMem+pc.totCost.Load() > pc.hardMemLimit.Load() {
+		return // do nothing if it exceeds the hard limit
+	}
+	headNode := pc.getHead(key, true)
```

Comment:
Looks like we can have a nil check here since `getHead` doesn't make sure it won't be `nil`.